### PR TITLE
chore: record event gql mutation

### PR DESCRIPTION
**Pull request by warrensplayer**
- Closes https://github.com/cypress-io/cypress-internal/issues/632

### Additional details

| NOTE | 
| -- |
| This PR is dependent on the Cloud PR: https://github.com/cypress-io/cypress-services/pull/6165  |

Adding additional way to capture event data in Cypress Cloud using a GQL mutation rather than separate endpoints.  This allows for recording different events with different payloads without continuing to expand the fields on the existing event types.

This PR introduces one new event called LocalTestCounts.  This passes 4 count values for specs and tests within a project as well as breaking out separate counts for files that match the example e2e paths.

NOTE: The count for tests is a rudimentary algorithm that is just looking for a regex in the files.  It is not meant to be totally accurate for when tests are created through functions, but rather give an estimate of the amount of effort put into writing tests.

#### How event is triggered

The `EventCollectorSource` watches two events that are triggered by other parts of the code (`authChange` and `configChange`).  I added two methods to subscribe (`subscribeToRawEvent`) and unsubscribe(`unsubscribeToRawEvent`) to those events directly in the `DataEmitterActions` without having to deal with the `AsyncIterator` that the GQL subscriptions need to operate.  

### Steps to test

To monitor the outgoing events, you can turn on DEBUG statements for:
* `cypress:data-context:sources:EventCollectorSource`
* `cypress:data-context:actions:EventCollectorActions`

```
DEBUG=cypress:data-context:sources:EventCollectorSource,cypress:data-context:actions:EventCollectorActions yarn dev
```

Preconditions that must be met for LocalTestCounts event to be sent:
* must be logged in
* must have loaded a project config
* must not have sent an event for this project within the last 30 days
 * this being tracked through a new saved setting in project state called `lastTestCountsEvent`.  This is saved as a timestamp for when the last event was sent

Scenarios to test:
* Project does not have `lastTestCountsEvent` in the `state.json` file for the project or the saved timestamp is older than 30 days
  * Local project has no spec files
    * Should see counts of all zeros
  * Local project is not connected to a cloud project
    * Should NOT see a `projectSlug` value in the event
  * Project has specs matching the path and name for the example scaffolded e2e specs 
    * Can use the `cypress-example-kitchensink` project as a test since that is where all the examples come from or scaffold a new project
      * Expect for the `example` counts for specs and tests to match the `total` counts
* Project has a  `lastTestCountsEvent` timestamp saved in the `state.json` file that is less than 30 days old
  * Should see a debug message from `cypress:data-context:sources:EventCollectorSource` that says "will not send" and shows that "hasBeenSentLast30Days" is `false`

Example steps of when to see the events be triggered:
* In global mode
  * after selecting a project and then a testing type, this will trigger the `configChange` event
* in regular project mode
  * after opening the App and selecting a testing type, this will trigger the `configChange` event
* in either mode
  * log out of the cloud
  * open a project and choose a testing type
  * log in, this will trigger the 'authChange' event   

### How has the user experience changed?

N/A

### PR Tasks

- [ ] Have tests been added/updated?
- [ ] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? 
- [ ] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?

### CODE REVIEW

File: packages/data-context/src/data/ProjectLifecycleManager.ts [hunk 1]
```diff
@@ -433,6 +433,11 @@ export class ProjectLifecycleManager {
         },
       })
 
+      s.eventCollectorSource?.destroy()
+      if (!this.ctx.isRunMode) {
```
Comment (lmiller1990) [inline packages/data-context/src/data/ProjectLifecycleManager.ts]: Going to flip this to `isOpenMode`

File: packages/data-context/src/sources/CloudDataSource.ts [hunk 2]
```diff
@@ -100,10 +103,13 @@ export class CloudDataSource {
   }
 
   get #additionalHeaders () {
-    return {
-      'Authorization': this.#user ? `bearer ${this.#user.authToken}` : '',
-      'x-cypress-version': pkg.version,
-    }
+    return (async () => {
```
Comment (lmiller1990) [inline packages/data-context/src/sources/CloudDataSource.ts]: An `async` getter seems unidiomatic, I think it's much more natural to use `await` with a function returning a `Promise`. I changed this to be a function.

File: packages/data-context/src/sources/EventCollectorSource.ts [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+  }
+
+  async sendLocalTestCounts () {
+    debug('Checking to send local test counts')
+    const user = this.ctx.coreData.user
+    const isAuthenticated = !!user && !!user.name
+    const projectSlug = await this.ctx.project.projectId()
+    let testingType: CloudTestingTypeEnum | undefined
+
+    switch (this.ctx.coreData.currentTestingType) {
```
Comment (lmiller1990) [inline packages/data-context/src/sources/EventCollectorSource.ts]: We should just return early here if there is no `currentTestingType`, although I don't think that is even possible?

**Review by lmiller1990 — COMMENTED** [review]
Still testing this

**Review by lmiller1990 — APPROVED** [review]
I pushed a few commits. This seem to work. I got these logs:

```
2023-08-01 15:32:40.403 Cypress[18355:103177] +[CATransaction synchronize] called within transaction
[cy:open:dev:18341]:   cypress:data-context:sources:EventCollectorSource Starting +0ms
[cy:open:dev:18341]:   cypress:data-context:sources:EventCollectorSource Checking to send local test counts +2s
[cy:open:dev:18341]:   cypress:data-context:sources:EventCollectorSource will not send - not authenticated +1ms
[cy:open:dev:18341]:   cypress:data-context:sources:EventCollectorSource Checking to send local test counts +10s
[cy:open:dev:18341]:   cypress:data-context:sources:EventCollectorSource will not send - not authenticated +1ms
[cy:open:dev:18341]:   cypress:data-context:sources:EventCollectorSource Checking to send local test counts +17s
[cy:open:dev:18341]:   cypress:data-context:sources:EventCollectorSource sending recordEvent for local test counts {
[cy:open:dev:18341]:   projectSlug: '7p5uce',
[cy:open:dev:18341]:   testingType: 'COMPONENT',
[cy:open:dev:18341]:   totalSpecs: 20,
[cy:open:dev:18341]:   totalTests: 21,
[cy:open:dev:18341]:   exampleSpecs: 0,
[cy:open:dev:18341]:   exampleTests: 0,
[cy:open:dev:18341]:   branch: 'wip-optimizing'
[cy:open:dev:18341]: } +12ms
[cy:open:dev:18341]: Invalid field: The field `cloudRecordEvent` does not exist on `Mutation`, but the GraphQL document expects it to exist.
[cy:open:dev:18341]: Traversal will continue, however this may lead to undefined behavior!
[cy:open:dev:18341]: (Caused At: "EventCollectorActions_RecordEvent" mutation)
[cy:open:dev:18341]: https://bit.ly/2XbVrpR#4
[cy:open:dev:18341]:   cypress:data-context:sources:EventCollectorSource result {
[cy:open:dev:18341]:   data: { cloudRecordEvent: true },
[cy:open:dev:18341]:   error: undefined,
[cy:open:dev:18341]:   extensions: undefined,
[cy:open:dev:18341]:   operation: {
[cy:open:dev:18341]:     key: 3665765446,
[cy:open:dev:18341]:     query: { kind: 'Document', definitions: [Array], loc: [Object] },
[cy:open:dev:18341]:     variables: { localTestCounts: [Object] },
[cy:open:dev:18341]:     kind: 'mutation',
[cy:open:dev:18341]:     context: {
[cy:open:dev:18341]:       _instance: [],
[cy:open:dev:18341]:       url: 'https://cloud.cypress.io/test-runner-graphql',
[cy:open:dev:18341]:       fetchOptions: undefined,
[cy:open:dev:18341]:       fetch: [AsyncFunction: fetch],
[cy:open:dev:18341]:       preferGetMethod: false,
[cy:open:dev:18341]:       suspense: false,
[cy:open:dev:18341]:       requestPolicy: 'cache-first'
[cy:open:dev:18341]:     }
[cy:open:dev:18341]:   },
[cy:open:dev:18341]:   errors: undefined
[cy:open:dev:18341]: } +976ms
[cy:open:dev:18341]:   cypress:data-context:sources:EventCollectorSource Checking to send local test counts +3s
[cy:open:dev:18341]:   cypress:data-context:sources:EventCollectorSource will not send { isAuthenticated: true, hasBeenSentLast30Days: true } +3ms
```

- I don't have the latest cloud checked out. I think this explains the `Invalid field: The field `cloudRecordEvent` does not exist on `Mutation`, but the GraphQL document expects it to exist.` - this feature has not been deployed yet.
- The counts look correct

**Review by astone123 — APPROVED** [review]
This looks good to me. I tested it out locally and saw the logs with the spec data getting sent, and I also confirmed that the data was not sent if I was logged out or if I opened the same project again that had already sent in the last 30 days

**Pull request merged**